js/eventloop/eventloop.go

```diff
@@ -55,11 +55,69 @@ func (e *EventLoop) wakeup() {
+// some asynchronous work off the main thread and that you may need to execute
+// some code back on the main thread when you are done. So, once you call this
+// method, the event loop will wait for you to finish and give it the callback
+// it needs to run back on the main thread before it can end the whole current
+// script iteration.
+//
+// ReservePendingCallback() *must* be called from the main runtime thread, but
+// its result scheduleOnMainThread() is thread-safe and can be called from any
+// goroutine. scheduleOnMainThread() ensures that its callback parameter is
+// added to the VU runtime's tasks queue, to be executed on the main runtime
+// thread eventually, when the VU is done with the other tasks before it. Unless
+// the whole event loop has been stopped, invoking scheduleOnMainThread() will
+// queue its argument and "wake up" the loop (if it was idle, but not stopped).
+//
+// Keep in mind that once you call ReservePendingCallback(), you *must* also
+// call scheduleOnMainThread() exactly once, even if don't actually need to run
+// any code on the main thread. If that's the case, you can pass an empty no-op
+// callback to it, but you must call it! The event loop will wait for the
+// scheduleOnMainThread() invocation and the k6 iteration won't finish and will
+// be stuck until the VU itself has been stopped (e.g. because the whole test or
+// scenario has ended). Any error returned by any callback on the main thread
+// will abort the current iteration and no further event loop callbacks will be
+// executed this iteration.
+//
+// A common pattern for async work is something like this:
+//
+//    func doAsyncWork(vu modules.VU) *goja.Promise {
+//        scheduleOnMainThread := vu.ReservePendingCallback()
+//        p, resolve, reject := vu.Runtime().NewPromise()
+//
+//        // Do the actual async work in a new independent goroutine, but make
+//        // sure that the Promise resolution is done on the main thread:
+//        go func() {
+//            // Also make sure to abort early if the context is cancelled, so
+//            // the VU is not stuck when the scenario ends or Ctrl+C is used:
+//            result, err := doTheActualAsyncWork(vu.Context())
+//            scheduleOnMainThread(func() error {
+//                if err != nil {
+//                    reject(err)
+//                } else {
+//                    resolve(result)
+//                }
+//                return nil  // do not abort the iteration
+//            })
+//        }()
+//
+//        return p
+//    }
+//
+// This ensures that the actual work happens asynchronously, while the Promise
+// is immediately returned and the main thread resumes execution. It also
+// ensures that the Promise resolution happens safely back on the main thread
+// once the async work is done, as required by goja and all other JS runtimes.
+func (e *EventLoop) ReservePendingCallback() (scheduleOnMainThread func(callback func() error)) {
```

Comment:
```suggestion
func (e *EventLoop) ReservePendingCallback() (queueCallback func(callback func() error)) {
```

`MainThread` seems:
1. strange   as it's a goroutine and we have multiple ones per each VU. Now arguably `schelduleOnMainThread := vu.ReservePendingCallback()` looks somewhat okay ... but:
2. arguably is overly long without too much additional info ... IMO - which again probably isn't the most important one in this particular topic :rofl: 

`schedule` IMO have more of a connotation we will "schedule it at some particular time". Also we do queue it in a queue ;)

Reply:
:+1: :+1: for `queueCallback`, will change it

Reply:
Resolved by https://github.com/grafana/k6/pull/2615/commits/8da5202aee2e0d0afd4a8f82fc5c0f01d2b8a339